tokeniser.go

```diff
@@ -57,70 +48,73 @@ func (t *Tokeniser) endWord() {
 		return
 	}
 
-	// This ensures any non-UTF8 is replaced with the Unicode replacement
-	// character.  We could use String(), but this would permit invalid
-	// UTF8.
-	uword := []rune{}
-	for {
-		r, _, err := t.word.ReadRune()
-		if err != nil {
-			break
-		}
-		uword = append(uword, r)
-	}
-
-	t.words = append(t.words, string(uword))
-	t.word.Truncate(0)
+	t.words = append(t.words, string(t.word))
+	t.word = []byte{}
 	t.inWord = false
 }
 
-// Tokenise feeds raw bytes into a Tokeniser.
+// Tokenise reads a tokenised line from the Reader.
 //
-// If the bytes include the ending of one or more command lines, those lines
-// shall be returned, as a slice of lines represented as slices of
-// word-strings.  Else, the slice shall be empty.
-//
-// Tokenise may return an error if its current word gets over-full.  In this
-// case, lines will contain the lines it managed to tokenise before keeling
-// over, count will contain the number of bytes processed (including the byte
-// causing the error), and the tokeniser will remain in error until replaced or
-// the current word is ended.
-func (t *Tokeniser) Tokenise(data []byte) (lines [][]string, count uint64, err error) {
-	count = 0
-	for _, b := range data {
-		// Exit early if the current word has become over-full.  This
-		// lets the caller handle the error without silently masking it
-		// if we manage to progress.
-		if t.err != nil {
-			break
-		}
+// Tokenise may return an error if the Reader chokes.
+func (t *Tokeniser) Tokenise() ([]string, error) {
+	// As per http://grokbase.com/t/gg/golang-nuts/139fgmycba
+	var bs [1]byte
 
-		if t.escapeNextChar {
-			t.put(b)
-			t.escapeNextChar = false
+	for {
+		// Constantly grab one byte out of the Reader.
+		// Technically inefficient, but this will be done on network
+		// connections mainly anyway, so this shouldn't be the
+		// bottleneck.
+		n, err := t.reader.Read(bs[:])
+		if err != nil {
+			return []string{}, err
+		}
+		// Spin until we get a byte.
+		if n == 0 {
 			continue
 		}
 
-		switch t.currentQuoteType {
-		case none:
-			t.tokeniseNoQuotes(b)
-		case single:
-			t.tokeniseSingleQuotes(b)
-		case double:
-			t.tokeniseDoubleQuotes(b)
+		lineDone := t.tokeniseByte(bs[0])
+		// Have we finished a line?
+		// If so, clean up for another tokenising, and return it.
+		if lineDone {
+			line := t.words
+			t.words = []string{}
+			return line, nil
 		}
+	}
+}
+
+// tokeniseByte tokenises a single byte.
+// It returns true if we've finished a line, which can only occur outside of
+// quotes
+func (t *Tokeniser) tokeniseByte(b byte) (endLine bool) {
+	endLine = false
```

Comment:
not a huge fan of the named return parameter in this case (or tokeniseNoQuotes) - just seems to confuse things